### ISSUE TITLE
[Feature/Meson]: make OpenGL optional, not mandatory

### ISSUE THREAD

**Issue by SupervisedThinking** [dosbox-staging/dosbox-staging#930]
**Is your feature request related to a problem? Please describe.**
If you don't disable OpenGL it will fail but since OpenGL is not mandatory to run dosbox it should be an optional dependency

```
Run-time dependency gl found: NO (tried pkgconfig and system)
Message: You can disable this dependency with: -Duse_opengl=false

../meson.build:178:2: ERROR: Dependency "gl" not found, tried pkgconfig and system
```
**Describe the solution you'd like**
It should not be necessary to set ` -D use_opengl=false` for OpenGL ES based distributions like most SBC ones.

**Additional context**
Buildlog:
```
UNPACK      dosbox-staging
BUILD      dosbox-staging (target)
    TOOLCHAIN      meson (auto-detect)
Executing (target): meson --prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --sysconfdir=/etc --libdir=/usr/lib --libexecdir=/usr/lib --localstatedir=/var --buildtype=plain -Dstrip=true --cross-file=/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/build/dosbox-staging-2c969cc12ae69163ee0afee77965f28641e18cc1/.armv8a-libreelec-linux-gnueabihf/meson.conf -D unit_tests=disabled /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/build/dosbox-staging-2c969cc12ae69163ee0afee77965f28641e18cc1
DEPRECATION: c_args in the [properties] section of the machine file is deprecated, use the [built-in options] section.
DEPRECATION: c_link_args in the [properties] section of the machine file is deprecated, use the [built-in options] section.
DEPRECATION: cpp_args in the [properties] section of the machine file is deprecated, use the [built-in options] section.
DEPRECATION: cpp_link_args in the [properties] section of the machine file is deprecated, use the [built-in options] section.
Using 'PKG_CONFIG_PATH' from environment with value: ''
The Meson build system
Version: 0.56.0
Source dir: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/build/dosbox-staging-2c969cc12ae69163ee0afee77965f28641e18cc1
Build dir: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/build/dosbox-staging-2c969cc12ae69163ee0afee77965f28641e18cc1/.armv8a-libreelec-linux-gnueabihf
Build type: cross build
Project name: dosbox-staging
Project version: 0.77.0
Using 'CFLAGS' from environment with value: '-march=native -O2 -Wall -pipe -I/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/include -Wno-format-security'
Using 'LDFLAGS' from environment with value: '-Wl,-rpath,/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/lib -L/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/lib'
Using 'CPPFLAGS' from environment with value: ''
C compiler for the host machine: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/armv8a-libreelec-linux-gnueabihf-gcc (gcc 10.2.0 "armv8a-libreelec-linux-gnueabihf-gcc-10.2.0 (GCC) 10.2.0")
C linker for the host machine: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/armv8a-libreelec-linux-gnueabihf-gcc ld.bfd 2.35.1
Using 'CXXFLAGS' from environment with value: '-march=native -O2 -Wall -pipe -I/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/include -Wno-format-security'
Using 'LDFLAGS' from environment with value: '-Wl,-rpath,/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/lib -L/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/lib'
Using 'CPPFLAGS' from environment with value: ''
C++ compiler for the host machine: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/armv8a-libreelec-linux-gnueabihf-g++ (gcc 10.2.0 "armv8a-libreelec-linux-gnueabihf-g++-10.2.0 (GCC) 10.2.0")
C++ linker for the host machine: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/armv8a-libreelec-linux-gnueabihf-g++ ld.bfd 2.35.1
Using 'CC_FOR_BUILD' from environment with value: '/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/host-gcc'
C compiler for the build machine: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/host-gcc (gcc 9.3.0 "gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0")
C linker for the build machine: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/host-gcc ld.bfd 2.34
Using 'CXX_FOR_BUILD' from environment with value: '/home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/host-g++'
C++ compiler for the build machine: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/host-g++ (gcc 9.3.0 "g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0")
C++ linker for the build machine: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/host-g++ ld.bfd 2.34
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: arm
Target machine cpu: armv8-a
Compiler for C++ supports arguments -Wextra-semi: YES 
Checking for function "clock_gettime" : YES 
Checking for function "mprotect" : YES 
Checking for function "realpath" : YES 
Checking whether type "struct dirent" has member "d_type" : YES 
Has header "pwd.h" : YES 
Has header "strings.h" : YES 
Has header "netinet/in.h" : YES 
Has header "sys/socket.h" : YES 
Checking if "test for setpriority support" compiles: YES 
Checking if "test for __builtin_expect support" compiles: YES 
Found pkg-config: /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/toolchain/bin/pkg-config (0.29.2)
Using 'PKG_CONFIG_PATH' from environment with value: ''
Run-time dependency opusfile found: YES 0.12
Run-time dependency threads found: YES
Using 'PKG_CONFIG_PATH' from environment with value: ''
Run-time dependency sdl2 found: YES 2.0.15
Using 'PKG_CONFIG_PATH' from environment with value: ''
Run-time dependency sdl2_net found: YES 2.0.1
Run-time dependency gl found: NO (tried pkgconfig and system)
Message: You can disable this dependency with: -Duse_opengl=false

../meson.build:178:2: ERROR: Dependency "gl" not found, tried pkgconfig and system

A full log can be found at /home/supervisedthinking/development/LibreELEC-RR/build.LibreELEC-RK3399.arm-10.0-devel/build/dosbox-staging-2c969cc12ae69163ee0afee77965f28641e18cc1/.armv8a-libreelec-linux-gnueabihf/meson-logs/meson-log.txt
FAILURE: scripts/build dosbox-staging during configure_target (default)
```

**Comment by dreamer** [comment]
OpenGL is already optional… opt-out. And it's going to stay this way for the foreseeable future.

If it was mandatory (which I thought about doing, but that would cut-off too many users), then there would be no option at all to disable it (just like e.g. there's no option to disable POSIX threads or SDL2).

So the question is: *why is it an opt-out instead of being automatically turned on/off depending on the system*.

The gist of it is: disabling OpenGL has serious functionality consequences (glshaders do not work now, 3dfx emulation won't work in the future) - user building the emulator needs to know about it.

I might revise this after OpenGL ES support will be available, but as for now: nope, closing this issue.

---
*Longer, a bit rambly answer*

It already was being turned off automatically back when we used Autoconf, before Meson migration… and we needed to deal with following issues in result:

1. Users (usually less experienced ones) build without necessary dependencies, don't read the logs from configure script, and then:
2. Complain publically that DOSBox Staging does not support glshaders or it's broken - this makes it harder for us to find out if the report is legit or is it yet-another-false-positive-because-user-did-not-pay-attention-while-building. (Because of course, the most effective way of reporting bugs is complaining on a random forum and never, ever saying anything in official bugtracker… grrr…)
3. Similar problem affects *us, developers* as well - only after making OpenGL an opt-out we discovered that MSYS2 builds are happening without OpenGL (which would explain fair number of those complaints from (2) - they are from users who went out of their way to not use our official build process on Windows).
4. I fully expect that problem in (2) is going to happen to overzealous packagers as well (e.g. some might have a bright idea to disable OpenGL on macOS because holy Apple Co. told them so…)

**Comment by SupervisedThinking** [comment]
`OpenGL is already optional… opt-out.` well this is not exactly what I would call optional. If it's optional then meson would report that it's not found and move on but actually the build fails and forces me to set a meson opt even though this is not necessary from a technical point of view.

You're correct if it comes to GLSL shaders but then again it's just a matter build magic, i have to "sed hack" them out for my ARM builds anyway because otherwise dosbox would steadily nag about an unkown conf opt. I agree that in theory this could increase the support burden but let's be real... someone who is able to build dosbox from source is most likely able to read some docs & log files. Meson could also simply tell the user at build time that if OpenGL is disabled then GLSL shaders would not be available.

It's not a big deal for me since the opt-out is available but it's like the meson opt `-Dunit_tests=disabled` IMHO this should be an opt-in since only devs will ever need it and so IMHO at least building dosbox should not fail without GL ¯\_(ツ)_/¯

Well of course if GLES and/or 3dfx/Glide will be added &  support then this discussion would be superflous although technically you could still run dosbox without GL.

**Comment by kcgen** [comment]
One of the major goals of **./build.sh** was using the right defaults for a given platform. On ARM systems, that means using OpenGLES, via SDL Texture Rendering. 

If you look at RetroPie, Recalbox, Batocera, and Lakka: it's all GLES via framebuffer. None of them using  OpenGL backed by X11 or Wayland.

kms, fake-kms, vkms, v3d, etc.. ARM and the Pi boards are still thrashing about trying to get this right. But until then - you will see [reports like this](https://www.raspberrypi.org/forums/viewtopic.php?f=67&t=277297) from a user getting 6 FPS rendering a single OpenGL quad on their X11 Rpi4 desktop.

We don't want users falling into that hole - and I now catch myself quickly _Ctrl+C_'ing my ARM builds when I  _forget to disable OpenGL_.  This is because I've installed openSUSE on my Pi (to test Aarch64), which also includes desktop and OpenGL libraries and headers. 

So circa-2021, we actually want OpenGL off on ARM; we also don't want package integrators _including_ OpenGL -- because  that essentially guarantees terrible performance on ARM (as of today).

**Comment by SupervisedThinking** [comment]
>One of the major goals of ./build.sh was using the right defaults for a given platform. On ARM systems, that means using OpenGLES via SDL Texture Rendering.

>If you look at RetroPie, Recalbox, Batocera, and Lakka: it's all GLES via framebuffer. None of them are using OpenGL backed by X11 or Wayland.

This is the primary reason why I opened this issue https://github.com/dosbox-staging/dosbox-staging/issues/926 & why IMHO OpenGL should be autodetected but no opt-out.

>kms, fake-kms, vkms, v3d, etc.. ARM and the Pi boards are still thrashing about trying to get this right. But until then - you will see reports like this from an Rpi4 user getting 6 FPS rendering a single quad-texture using OpenGL on their X11 desktop.

Beside the vendor hacked versions of various distros most users will probably use Armbian or Raspbian *if they use a full size desktop distro* at all. The driver side is pretty clear nowadays... GLES / GBM & KMS which relies on Mesa 3D drivers like https://docs.mesa3d.org/drivers/vc4.html or https://docs.mesa3d.org/drivers/panfrost.html and technically there is no need for OpenGL on a desktop platform since you could run Wayland builds with GLES only.

>We don't want users falling into that hole - and I now catch myself Ctrl+C'ing my ARM builds when I forget to disable OpenGL. This is because I've installed openSUSE on my Pi (to test Aarch64), which also includes desktop and OpenGL libraries and headers.

I'm pretty sure the average user of a SBC has some common sense and does not expect outstandig performance. Yes OpenGL headers are always installed by Mesa 3D, for whatever reason, & the RPi3B could support OpenGL 2.1 which is rather... dated.

>So circa-2021, we actually want OpenGL off on ARM; we also don't want package integrators including OpenGL -- because that essentially guarantees terrible performance on ARM (as of today).

From my perspective typical package maintainers are aware of such problems & I don't see how keeping OpenGL opt-out would help in this case, if you're running a desktop distro with X11 & GL & if you build dosbox from source it'll pull in GL and... meh

>Given OpenGL is a pessimistic setting for ARM across the board circa 2021 (versus GLES), it would be great if OpenGL was default-disabled just on ARM, but could still be turned on if desired.

IIRC OpenGL comes with X11 or rather GLX which probably most stuff uses. It's there for compat reasons but every sane distro would make it's switch to Wayland + GLES, if you look at libretro and else most of them already have GLES renderer support.

E.g. LibreELEC can be build with a Wayland/Weston displaymanger too & ditch X11 completely which wasn't done till today because Nvidia has to pull out his head of his a***

**Comment by kcgen** [comment]
@SupervisedThinking  - I'm not sure if you're for or against what I'm suggesting. 

If we leave ARM as-is, the most "default release" builds _are going to get opengl_ (like you said - libraries and headers are often included)  

The default `conf` settings plus the logic in the code (when OpenGL was enabled) will Staging prefer the slow `output = opengl` route instead of `output = texture` + SDL2's GLES backend.

**Comment by dreamer** [comment]
> Given OpenGL is a pessimistic setting for ARM across the board circa 2021 (versus GLES), it would be great if OpenGL was default-disabled just on ARM, but could still be turned on if desired.

OpenGL support has nothing to do with CPU architecture - we are not going to disable OpenGL based on that.

@kcgen @SupervisedThinking 
Perhaps what you really need is a file where you can define the desired set of features for specific devices?

In Meson this can be accomplished by using [machine files](https://mesonbuild.com/Machine-files.html). We already use them for configuring e.g. Clang builds on CI.

That means we would have a new `ini` file, e.g. `linux-arm-libreelec-rk3399.ini`, in that file we can specify all the details how to accomplish optimal build on that specific device (what compiler should be used, what flags should be passed, etc) - and additionally what features should be enabled ([project-specific-options](https://mesonbuild.com/Machine-files.html#project-specific-options)):
```ini
[project options]
use_opengl = false
```
Meson setup step would look like this:
```
meson setup --native-file=contrib/machine/libreelec-rk3399.ini
```
@kcgen you could simply make it an alias on your machine:
```sh
alias meson-setup=meson setup --native-file=contrib/machine/raspberry-pi-3b.ini
```
Is this something that would solve both of your problems? :)

**Comment by SupervisedThinking** [comment]
@kcgen IMHO the meson build just should not fail if OpenGL is not present. If some maintainer builds dosbox with OpenGL for a SBC distro & if some user runs dosbox with `output = opengl` well c'est la vie. It's up to the user to configure his conf files properly if you ask me.
Beside that... if you could adopt the OpenGLES patches I've linked... well that would be perfect ^^

@dreamer what creates the default dosbox conf files? I clean up the conf files this way which is fine for now

https://github.com/SupervisedThinking/LibreELEC-RR/blob/libreelec-10.0-RR/packages/supervisedthinking/emulation/standalone/dosbox/dosbox-staging/package.mk#L48-L53

In the end it boils down to `ARM=GLES` and you don't really need machine specific files... afaik most, if not all, currently available ARM boards have at least GLES 2.0 support or even 3.x so nobody cares about OpenGL at all ^^

**Comment by kcgen** [comment]
> @kcgen IMHO the meson build just should not fail if OpenGL is not present.

Absolutely! I'm suggesting exactly that; we use some mechanism in meson to:
1. Disable OpenGL by default on ARM, but still ...
2. Leave it optional if users turn back-on `-Duse_opengl=true`.

> if some user runs dosbox with output = opengl well c'est la vie.

`output = auto` will, behind the scenes, use the opengl path_  (ie: equal to `output = opengl`) if meson found the openGL libraries and headers at link-time.  That's why ARM with OpenGL on is silently harmful.  The entire configuration is not what we want on ARM. 

> It's up to the user to configure his conf files properly if you ask me.

1. It's up to the **developers** to provide the best possible baseline configuration for a given platform
2. It's up to the **package maintainers** (or person compiling) to adjust it for their distribution
3. It's up to the **users** to adjust based on custom hardware, personal preferences, and games' needs

We should strive toward that, as far as we can go, given the tools at hand. 

@dreamer , I like your suggest of:

- `meson setup --native-file=contrib/machine/linux-libreelec-rk3399.ini`
- `meson-setup=meson setup --native-file=contrib/machine/linux-raspberry-pi-3b.ini`

How can these simply apply by default on the given platform?

I want `meson setup` to query the system, see I'm running on _any_ ARM system, and use the couple options we might set in `contrib/machine/linux-arm.ini`.

I want us to start at the most coarse level because, currently, there is only one common settings we care about for all ARM systems (which is disabling OpenGL by default).

If the time comes for more nuances settings.. saying between the Pi4 vs RockPro64 - then yeah, we can start breaking out more INI files.

**Comment by SupervisedThinking** [comment]
Well I'm not sure if you're overcomplicating things 😅

The overall performance on a SBC can extremely differ depending on the actual SoC, distribution, APIs, blob or foss driver, mainline or bsp kernel.

But this isn't IMHO not up to you and not your problem at all.

From my pov I would do three things:

- make opengl opt-out
-  add a notification at build that GLSL shader are depending on GL and not available if disabled
- add a notification on ARM + GL that the actual performance users can expect is quite low

So nobody can nag about anything because you're warned/informed about the current state. Also if I use the Amlogic BSP kernel and X11 blob then I can expect a way different performance than a RPi user with his mediocre SoC with awful GPU and WIP foss drivers.

**Comment by kcgen** [comment]
> From my pov I would do three things:  make opengl opt-in

Agreed. More specifically, on ARM-only.

Also agree with your other points :+1:, also on ARM. 

Will see what kind of meson magic @dreamer can suggest, to make meson do this for us :slightly_smiling_face:

**Comment by SupervisedThinking** [comment]
Well does the emulator run at full speed no matter what you OpenGL performance is once you set `output = texture`? If so then I don't see why it should not pull in GL automatically if it's available. Again if a user or package maintainer builds dosbox-staging from scratch **and does not read the meson opts/wiki/manual** ... well sorry but that's the problem of the dude who builds crap then.

Just because some guys at the raspberry pi forum have had bad experiences with their wip drivers does not mean all user would have these issues: https://github.com/rockchip-linux/libmali/tree/master/lib/aarch64-linux-gnu there are plenty of alternatives & your mileage may vary.

**Comment by kcgen** [comment]
Are there any common ARM SBCs where: `output = opengl` produces equal or better results than `output = texture` + SDL's GLES backend?

I'm specifically talking about DOSBox Staging.

My understanding is that GLES is (as of today) is always the better choice.

But if that's not the case, then yeah, I suppose we need fine-grained platform differentiation.

Certainly on all the current Pi* boards, disabling OpenGL in DOSBox Staging at compile-time is the right choice.

**Comment by SupervisedThinking** [comment]
It's simple. If you're running Retropie, Batocera, Lakka or something then the SBC most likely uses GBM/KMS and has no windowmanager at all. If someone is running a full sized desktop distribution then X11 is available anyway & so most likely OpenGL & GLX.

But as I said... IMHO it's up to the package maintainer or user to adjust the build flags to create a proper package. But the buildsystem on it's own should only pull in what it actually needs, no matter how the actual perfomance might be.

E.g. you can technically build mame2016 or mame for a RPi3 too, but it won't run decent. But you can build whatever you like 😄

**Comment by kcgen** [comment]
> the buildsystem should only pull in what is available, no matter how the actual performance might be

The beauty of a push-button build system is reduces the barrier for users to track closer to the the head of the source repo.  I was such a user for almost a decade with SVN, and still am for something like 12 other console emulators. 

If a common platform dependency is needed to ensure acceptable performance _and_ the user hasn't indicated they're aware of this, then we have a duty, as the more knowledgeable maintainer, to inform them (if possible).

The person building the source can then either provide the high-performance dependency or indicate they accept the consequences, with something like `"-Dcarry_on_with_busted_performance=true"`.

But I realize this is an area where we have different opinions. 

Is there an actionable improvement we can make here?

@dreamer suggested putting sane defaults in platform-specific files:

 - `contrib/machine/linux-libreelec-rk3399.ini`
 - `contrib/machine/linux-raspberry-pi-3b.ini`

I don't think either of us would object to this.

I would go a step further an ask that Meson be made to _use these by default on said platforms_, and not require the user create custom scripts, aliases, or other shell-side short-cuts to provide these.

In other words, if I'm on a Pi 3B, and run `meson setup -Dbuild-type=release build`, then I want whatever defaults are in `contrib/machine/linux-raspberry-pi-3b.ini`.

Is that also accepted to you, @SupervisedThinking ?

**Comment by SupervisedThinking** [comment]
If you need some chart to get an overview about (somewhat) recent SoCs and their performance:

https://github.com/ThomasKaiser/sbc-bench/blob/master/Results.md

And to have an imagination what is technically possible with an ARM SBC:
https://www.youtube.com/watch?v=Y8N6qZzWqic

So unless you want to add dozens of .ini files... just set OpenGL to autodetect and let it be. It would not change the current experience for users on ARM boards with GL & it would improve and ease build scripts for GLES builds.

For everything else... well IMHO the user is responsible. 

OT
>I get your point and I'm just a "pro-user"/maintainer myself. I have some technical knowledge but I'm not a programmer at all... I started about 3 years ago to fork LibreELEC to build a distro which suits my needs because I wanted a wholesome distro... something that runs my HTPC & also lets me or my family and else play some games. There was a fork around which I found in the LE forum which had... well several flaws from my perspective. So I've studied for weaks the github repo how to get things up and running and started to rebase a similar but better fork. I tend to write scripts for everything, set up basic confics that run well to ease usage for others who want to use my builds but I alway tell them **rtfm**! If someone wants to use FOSS and is to lazy to read how to use it... well sorry but not sorry and certainly **not my problem**.

**Comment by kcgen** [comment]
I understand what the hardware is capable of in general - when optimally programmed.  But we're talking about DOSBox Staging's limitations and inefficiencies, and what's best for it.

> just set OpenGL to autodetect and let it be

That's not acceptable when building DOSBox Staging on platforms where OpenGL _is_ required for acceptable performance.

However, _it is an acceptable default_ on ARM platforms.

> So unless you want to add dozens of .ini files

I don't believe that's needed.  In this case, we simply need Meson to detect that it's running on ARM and make OpenGL optional. @dreamer - can that be done?

**Comment by dreamer** [comment]
> Are there any common ARM SBCs where: `output = opengl` produces equal or better results than `output = texture` + SDL's GLES backend? Specifically in DOSBox Staging.

How can we differentiate between SBCs from other ARM-based machines? Machines based on Apple M1, NVIDIA Tegra, perhaps various Chromebooks as well are going to prefer `output = opengl` over `output = texture`.

> My understanding is that GLES is (as of today) is always the better choice.

Depends on what do you mean by that. Generally, I would agree, but, as it stands right now, glshaders do not work with `output = texture`. And, importantly, **there is no replacement for `glshader = sharp`** (no, `texturenb` is not even close).

If we could wave a magic wand and replace our old, crusty, bad code for `output = opengl` with OpenGL ES 3.0 - I would totally do it. But we can't.

I think it's time to start development on a new graphics pipeline that would replace all that bad code in `sdlmain.cpp`. And I mean total replacement - I have enough of all bugs and limitations caused by current graphics pipeline - there are so many of them:

- vsync
- problems with going in/out of fullscreen
- problems with splash screen on KDE
- problems with window initialization on Windows
- impossible to create nice overlays to give feedback to user
- broken window resizing on Windows and macOS
- … and now also people complaining about slow performance because they compiled with OpenGL support on Raspberry Pi…

Fixing these problems one-by-one in the code as it exists today is, frankly, borderline impossible. We can't use any of the patches that people used for OpenGLES before us, because they all just added more incompatible cruft and none of them actually focused on cleaning the existing crap (all of that is work is useless anway since upstream implemented glshaders) :disappointed: 

I started some experimentation with OpenGL ES 3.0, but this won't be ready for 0.77.0, nor for 0.78.0. I want new graphics pipeline to be designed in a way where we'll be able to relatively easy expand the support for WebGL and Vulkan in the future.

#### Now, back to this discussion - about making OpenGL automatic feature instead of being opt-out

@kcgen I will not agree to disabling OpenGL based on CPU architecture. Why would we want to automatically disable `glshaders` on M1 or on Tegra?

If we changed the buildsystem as @SupervisedThinking suggests - to make OpenGL enabled/disabled automatically based on the libraries available at compile time - would it solve your problem? Can you somehow get rid of OpenGL on your system, keep SDL2 and OpenGLES, and rely on feature being automatically disabled?

**Comment by kcgen** [comment]
> @kcgen I will not agree to disabling OpenGL based on CPU architecture. Why would we want to automatically disable glshaders on M1 or on Tegra?

I agree, and see the dillema of not being able to piece-meal out the SBC's versus first-class OpenGL ARM platforms like the M1. Argh.

**Comment by dreamer** [comment]
So, how about making OpenGL automatic instead of opt-in? As OP requested in this issue? Can this be a solution for your environment?

**Comment by kcgen** [comment]
It won't worsen things on the Pi* versus the existing configuration. 

That is, If Meson builds with OpenGL (because it's set to `auto`), the Pi users will be greated with equally bad performance they already get.

But on platforms where OpenGL is make-or-break when it comes to performance (ie: only fallback is software rendering), then yes - I think this will worsen things for those users.

**Comment by SupervisedThinking** [comment]
Well I'm pretty sure a M1 would not support OpenGL direct anyway but Metal faking some dated... GL 4.1 or something?

Beside that afaik you can have OpenGL & OpenGL ES in parallel on the same system. But it also depends on your windowserver, X11 will probably use GLX & Wayland would run EGL etc.

So in the end Pi users could just set `output = texture` and it would be fine? Beside that package maintainers should set sane flags and disable or mention the quirks around OpenGL on a RPi.

**Comment by SupervisedThinking** [comment]
> OpenGL and OpenGL ES runs fine side-by-side on desktop Linux distros (at least with Mesa), user does not need to do anything :)

Correct, well at least if the system is correct configured 

>As for M1: AFAIK OpenGL is still there… it's outdated - but OpenGL code we inherited from DOSBox is even more outdated :(

I thought it uses Metal as translation layer like Zink translates OpenGL for Vulkan ^^ anyway DOSBox has no outstanding 3D graphic, just some shaders so I guess dated GL is fine.

>It will hurt me, as I will deal with users complaining about broken shaders / spreading misinformation :(

As I said just spread the news at build time. Beside that everything will stay as it is, if you have no GL you won't get GL just because meson moves on.

I have 3 unified conf files, one for just booting the cli & two templates for FS & MT32. Both are based on Generic X86 configs & were adjusted for the ARM package builds. Dosbox itself would create an empty config if none is present but that's not what I need or want on an embedded system since it needs to mount specific paths for soundfonts or the path to DOS games.
